### DISCUSSION TITLE
Provide a separate package for typings

### DISCUSSION THREAD

**Discussion by WoLfulus** [directus/directus#3025] [Feature Requests]
Writing extensions in typescript isn't that cool without properly using types. It would be nice to have something like `npm install @directus/support` so we can import the abstract types and interfaces to enhance the writing experience and reliability of endpoints, hooks and anything we might want to expose.

Btw: `support` is the name used in Laravel for the abstract types the framework exposes so that packages can depend on them without creating a hard dependency on the framework itself (helps with versioning and stuff).

```js
module.exports = function registerEndpoints(router, context) {
  const database = context.db; // Passes
};
```

vs

```ts
import { Router, EndpointContext } from '@directus/support';

export default function registerEndpoints(router: Router, context: EndpointContext): void {
  const database = context.db; // Error missing `db`
}
```

**Comment by rijkvanzanten** [comment]
Wouldn't you be able to pull them from the `directus` package directly? 🤔 

We should be able to tell TS to build a `index.d.ts` on build of the API, after which you should be able to pull in all the types directly from the source

**Reply by WoLfulus** [reply]
How would it work when we start publishing extensions to NPM for example? For example, I'm using Directus `v9.2.0`, extension was written with a dependency for `v9.0.5` in which nothing has changes from the imported types. 

Also I couldn't import any types from the current `directus` package. At least it doesn't show anything. Is that already working? Can we generate the index.d.ts for  beta.11?

**Reply by rijkvanzanten** [reply]
> How would it work when we start publishing extensions to NPM for example? For example, I'm using Directus v9.2.0, extension was written with a dependency for v9.0.5 in which nothing has changes from the imported types.

that should be alright, as you've built and bundled the extensions for an older version. As long as there aren't any breaking changes in between those two versions, the extension will work as expected as it's only running the built version

> Also I couldn't import any types from the current directus package. At least it doesn't show anything. Is that already working? Can we generate the index.d.ts for beta.11?

No this is not a thing yet, but i'm saying we could / should?

**Comment by apstanisic** [comment]
I ran into the same problem. As a workaround I cloned this repo inside my project, and created context type.
Router type I just imported from `@types/express`

```typescript
import * as services from "../next/api/src/services";
import * as exceptions from "../next/api/src/exceptions";
import database from "../next/api/src/database";
import env from "../next/api/src/env";

export type EndpointContext = {
  services: typeof services;
  exceptions: typeof exceptions;
  env: typeof env;
  database: typeof database;
};
```

**Comment by nickluger** [comment]
Workaround:
```ts
// Get types directly, where available
import { ExtensionContext } from "directus/dist/types";
// Built your own types, see below ⬇️
import { HookEventHandlerContext } from "../../types";

export default function registerHook(context: ExtensionContext) {
  return {
    "items.update": function (eventContext: HookEventHandlerContext) {
        // do stuff
    },
  };
}
```
```ts
// your own types built from existing types (i did not find this type directly, maybe it exists somewhere...)
import {
  Accountability,
  PrimaryKey,
  SchemaOverview,
} from "directus/dist/types";

export interface HookEventHandlerContext {
  event: string;
  accountability: Accountability;
  collection: string;
  item: PrimaryKey;
  action: string;
  payload: Record<string, any>;
  schema: SchemaOverview;
}

```

**Comment by johnhuffsmith** [comment]
Adding to this conversation; I'm finding that I'm having to recreate types in my extension as well.  Specifically `ApiExtensionContext` defined in `@directus/shared` is full of types that are defined in the `api` project; but `@directus/shared` is pretending like it's not dependent on those `api` types

Options:
1. make `api` a dependency of `@directus/shared` (ew)
2. move the the missing types into `@directus/shared` (meh)
3. make a new `@directus/types` package to consolidate all the types that are shared among directus libs?

I think resolving these extension typing issues would be a huge win for developer ergonomics. I'd be happy to implement one of the options with some buy-in from core devs.

**Reply by joselcvarela** [reply]
There's a third option which is making a PR to https://github.com/DefinitelyTyped/DefinitelyTyped but I am sure who don't want to go that way.

Like WoLfulus mentioned, while developing npm extensions it does not feel alright to have to include all Directus to just have typings. So I vote for your third option.

Although, this will be a massive refactor, since we are using types everywhere and that makes me a little nervous 😨

**Reply by nickrum** [reply]
> make a new `@directus/types` package to consolidate all the types that are shared among directus libs?

That's precisely what `@directus/shared/types` is for. We opted for having a single package for any code shared between directus packages.

> Although, this will be a massive refactor, since we are using types everywhere and that makes me a little nervous 😨

This shouldn't be _too_ bad, actually. A lot of our types are already in `@directus/shared/types` 😉

My plan was to add interfaces in `@directus/shared/types` for `Emitter`, all of the exceptions and all of the services and let the classes in `directus` implement them. In addition to that, something similar has to be done for the App's stores. I just haven't had the time to do this yet.
@johnhuffsmith let us know, what you think. We'd glady review and merge a PR implementing this.

**Reply by johnhuffsmith** [reply]
Great! I'll see if I can move those dependent types into @directus/shared/types

**Reply by johnhuffsmith** [reply]
@nickrum @joselcvarela for your consideration: https://github.com/directus/directus/pull/11179

**Reply by johnhuffsmith** [reply]
@nickrum I'm realizing now what you meant by: 

> My plan was to add interfaces in @directus/shared/types for Emitter

and that the changes I made for `SchemaOverview` wont really work for the `services`, `exceptions`, or `emitter` props since it would involve ripping out the guts of `api` and moving them into `shared`.  It also doesn't seem reasonable to duplicate all the types in shared.  It seems like the root of the issue is that custom hooks depend on `api` philosophically; so we could either add `api` as a dependency somewhere (maybe on `extension-sdk`), or try to move some of the extension code into `api`. 

Maybe there's some typescript magic to make this all work together?
